rdf-abac-core/src/test/java/io/telicent/jena/abac/TestCachedAttributeStore.java

```diff
@@ -0,0 +1,267 @@
+        //then
+        assertEquals(initialResult, subsequentResult);
+        verify(mockedStore, times(1)).getHierarchy(attribute);
+    }
+
+    @Test
+    public void test_hasHierarchy_handleNullResult_asFalse() {
+        // given
+        AttributesStore mockedStore = Mockito.mock(AttributesStore.class);
+        Attribute attribute = new Attribute("Test");
+        when(mockedStore.getHierarchy(attribute)).thenReturn(null);
+        AttributeStoreCache cut = new AttributeStoreCache(mockedStore, Duration.ofSeconds(10), Duration.ofSeconds(10));
+        // when
+        boolean result = cut.hasHierarchy(attribute);
+        //then
+        assertFalse(result);
+        verify(mockedStore, times(1)).getHierarchy(attribute);
+    }
+
+    @Test
+    public void test_hasHierarchy_handleEmptyResult_asFalse() {
+        // given
+        AttributesStore mockedStore = Mockito.mock(AttributesStore.class);
+        Attribute attribute = new Attribute("Test");
+        Hierarchy emptyHierarchy = Hierarchy.create(attribute, emptyList());
+        when(mockedStore.getHierarchy(attribute)).thenReturn(emptyHierarchy);
+        AttributeStoreCache cut = new AttributeStoreCache(mockedStore, Duration.ofSeconds(10), Duration.ofSeconds(10));
+        // when
+        boolean result = cut.hasHierarchy(attribute);
+        //then
+        assertFalse(result);
+        verify(mockedStore, times(1)).getHierarchy(attribute);
+    }
+
+    @Test
+    public void test_users_callGoesToUnderlyingStoreNotCache() {
+        //given
+        AttributesStore mockedStore = Mockito.mock(AttributesStore.class);
+        when(mockedStore.users()).thenReturn(Set.of("user1", "user2"))
+                                 .thenReturn(Set.of("user1", "user2", "user3", "user4"));
+        AttributeStoreCache cut = new AttributeStoreCache(mockedStore, Duration.ofSeconds(10), Duration.ofSeconds(10));
+        //when
+        Set<String> initialResult = cut.users();
+        Set<String> subsequentResult = cut.users();
+        //then
+        assertNotEquals(initialResult, subsequentResult);
+        verify(mockedStore, times(2)).users();
+    }
+
+    @Test
+    public void test_attributes_initialCallMadeToUnderlyingStoreAfterCacheExpiry() {
+        //given
+        AttributesStore mockedStore = Mockito.mock(AttributesStore.class);
+        when(mockedStore.attributes("user")).thenReturn(SAMPLE_ATTRIBUTE_VALUE_SET)
+                                            .thenReturn(AttributeValueSet.of("attribute1=value1"));
+        AttributeStoreCache cut = new AttributeStoreCache(mockedStore, Duration.ofNanos(1), Duration.ofNanos(1));
+        //when
+        AttributeValueSet initialResult = cut.attributes("user");
+        AttributeValueSet subsequentResult = cut.attributes("user");
+        //then
+        assertNotEquals(initialResult, subsequentResult);
+        verify(mockedStore, times(2)).attributes("user");
+    }
+
+    @Test
+    public void test_put_overwritesUnderlying() {
```

Comment:
Do we actually want the cache to support writes at all, especially if it can overwrite the underlying store, that seems like a code smell?